crates/valence_client/src/lib.rs

```diff
@@ -960,6 +1006,62 @@ fn read_data_in_old_view(
     );
 }
 
+fn update_layer_view(
+    mut clients: Query<(
+        &mut Client,
+        &mut EntityRemoveBuf,
+        &OldPosition,
+        &OldViewDistance,
+        Option<&ClientLayerSet>,
+    )>,
+    entities: Query<(EntityInitQuery, &OldPosition, Option<&Layer>)>,
+) {
+    clients.par_iter_mut().for_each_mut(
+        |(mut client, mut remove_buf, old_pos, old_view_dist, client_layer_set)| {
+            // TODO: cache the chunk position?
+            let old_chunk_pos = old_pos.chunk_pos();
+
+            let view = ChunkView::new(old_chunk_pos, old_view_dist.0);
+
+            // Send entity spawn packets for entities that are in an entered layer and
+            // already in the client's view.
+            for (entity, &old_pos, layer) in entities.iter() {
+                if view.contains(old_pos.chunk_pos()) {
+                    if let (Some(client_layer_set), Some(layer)) = (client_layer_set, layer) {
+                        if client_layer_set
+                            .added()
+                            .collect::<HashSet<&LayerType>>()
+                            .contains(&layer.0)
+                        {
+                            entity.write_init_packets(old_pos.get(), &mut client.enc);
+                        }
+                    }
+                }
+            }
+
+            // Send entity despawn packets for entities that are in an exited layer and
+            // still in the client's view.
+            for (entity_init_item, &old_pos, layer) in entities.iter() {
+                if view.contains(old_pos.chunk_pos()) {
+                    if let (Some(client_layer_set), Some(layer)) = (client_layer_set, layer) {
+                        // if client_layer_mask.0.ne(&old_layer_mask.0) {debug!("client_layer_mask:
+                        // {:?}, old_client_layer_mask: {:?}, layer: {:?}", client_layer_mask,
+                        // old_client_layer_mask, layer);}
+                        if client_layer_set
+                            .removed()
+                            .collect::<HashSet<&LayerType>>()
+                            .contains(&layer.0)
```

Comment:
Maybe instead of collecting it into a `HashSet`, we use `find` to save memory allocation and need to read through all items returned by `added()`?
```rust
if client_layer_set
    .removed()
    .find(|x| **x == layer.0)
    .is_some()
```